### ISSUE TITLE
Longer prompts = CPU bottleneck = low GPU usage = lower generation speed

### ISSUE THREAD

**Issue by Zotikus1001** [Haoming02/sd-webui-forge-classic#130]
So, a while back I reported lower GPU usage when you were adding sage and cublas, and it seemed random. And we never found the cause.

Today I found what causes it, and its longer prompts. Past a certain length, it always happens.

There is some kind of CPU bottleneck, maybe because of the splitting/token length thing?

Note, it's only apparent if your GPU is fast enough and/or your CPU is slow enough (or your gen settings/resolution is low enough) to get bottlenecked, since in hiresfix of 1.5x it never happens.

I tried limiting Maximum Processor State in windows power settings, and the CPU bottleneck becomes more apparent. The slower the CPU is, the worse the gen speed is.

![Image](https://github.com/user-attachments/assets/25b576e1-7278-4bdf-8668-9f8a8c4e5c7c)

Low GPU usage:

> A sprawling, rain-slick cyberpunk alleyway stretching deep into the distance, its uneven cobblestone and cracked asphalt glistening under countless reflections from neon signs in Japanese, Korean, and English. The air is heavy with mist and steam rising from rusted sewer grates, curling into the night. On both sides of the narrow street, towering buildings loom oppressively high, their surfaces patched with layers of faded graffiti, peeling posters, and tangled masses of electrical cables snaking across facades. Massive digital billboards flicker and glitch with animated holographic advertisements, projecting soft light onto the wet ground in rippling hues of magenta, cyan, and amber.
> 
> The smell of rain mingles with sizzling oil from a tiny street food stall sheltered under a patched red awning, where a weary vendor in a cybernetic arm prosthetic ladles steaming broth into paper bowls. Beside him, a rusty vending machine hums, its touchscreen smudged and barely functional. Shadows shift in the distance, lone pedestrians wrapped in tattered trench coats, some with glowing visors, others with LED tattoos pulsing faintly beneath translucent rain hoods. Far above, through the narrow canyon of skyscrapers, faint aerial traffic drones cut through the night sky, their lights blinking in rhythmic patterns.
> 
> The entire scene is bathed in an atmospheric haze that diffuses light into soft halos, with occasional sparks from exposed wires crackling against the damp air. Puddles mirror the glowing chaos above, disturbed only by ripples from dripping eaves and the occasional scurrying rat. The mood is cinematic and immersive, evoking a sense of mystery, urban decay, and an unspoken tension just beneath the surface of the city’s hum.
> Negative prompt: bad quality
> 
> Steps: 40, Sampler: Euler a CFG++, Schedule type: Beta, CFG scale: 1.5, Seed: 200207587, Size: 896x1152, Model hash: ac80aef46e, Model: HiyokoDarkness_v20_VPRED, VAE hash: 235745af8d, VAE: sdxl_vae_fp16_fix.safetensors, Clip skip: 2, NGMS: 0.5, Noise Schedule: Zero Terminal SNR, Beta schedule alpha: 0.6, Beta schedule beta: 0.6, Version: classic
> 
> Time taken: 8.3 sec. A: 8.96 GB, R: 10.06 GB, Sys: 11.4/15.9922 GB (71.5%)

Normal GPU usage:

> A sprawling, rain-slick cyberpunk alleyway stretching deep into the distance, its uneven cobblestone and cracked asphalt glistening under countless reflections from neon signs in Japanese, Korean, and English. The air is heavy with mist and steam rising from rusted sewer grates, curling into the night. On both sides of the narrow street, towering buildings loom oppressively high, their surfaces patched with layers of faded graffiti, peeling posters, and tangled masses of electrical cables snaking across facades. Massive digital billboards flicker and glitch with animated holographic advertisements, projecting soft light onto the wet ground in rippling hues of magenta, cyan, and amber.
> 
> The smell of rain mingles with sizzling oil from a tiny street food stall sheltered under a patched red awning, where a weary vendor in a cybernetic arm prosthetic ladles steaming broth into paper bowls. Beside him, a rusty vending machine hums, its touchscreen smudged and barely functional. Shadows shift in the distance, lone pedestrians wrapped in tattered trench coats, some with glowing visors, others with LED tattoos pulsing faintly beneath translucent rain hoods. Far above, through the narrow canyon of skyscrapers, faint aerial traffic drones cut through the night sky, their lights blinking in rhythmic patterns.
> Negative prompt: bad quality
> 
> Steps: 40, Sampler: Euler a CFG++, Schedule type: Beta, CFG scale: 1.5, Seed: 200207587, Size: 896x1152, Model hash: ac80aef46e, Model: HiyokoDarkness_v20_VPRED, VAE hash: 235745af8d, VAE: sdxl_vae_fp16_fix.safetensors, Clip skip: 2, NGMS: 0.5, Noise Schedule: Zero Terminal SNR, Beta schedule alpha: 0.6, Beta schedule beta: 0.6, Version: classic
> 
> Time taken: 6.2 sec. A: 8.96 GB, R: 10.06 GB, Sys: 10.5/15.9922 GB (65.4%)

Only thing I did was reduce the prompt size a bit.

Is this something we can't avoid with long prompts? or is there something weird going on?

My cpu: AMD Ryzen 9 7950X3D, so a pretty good one :)
I also forced python.exe to run only on the CCD1 (The non-3D cache cores). I tried CCD0 only, all cores, and CCD1 only. And it seems it's even slower on 3D cache cores or all cores.

**Comment by Haoming02** [comment]
<h3 align="center">Doing 3 runs after 1 warm-up run</h3>

- **Long Long Prompt:**
```
Time taken: 4.1 sec. A: 7.42 GB, R: 7.88 GB, Sys: 9.2/15.9922 GB (57.6%)

Time taken: 4.0 sec. A: 7.43 GB, R: 7.88 GB, Sys: 8.6/15.9922 GB (53.6%)

Time taken: 4.0 sec. A: 7.43 GB, R: 7.88 GB, Sys: 8.7/15.9922 GB (54.4%)
```

- **Long Prompt:**
```
Time taken: 3.7 sec. A: 7.43 GB, R: 7.94 GB, Sys: 9.3/15.9922 GB (57.9%)

Time taken: 3.7 sec. A: 7.43 GB, R: 7.94 GB, Sys: 8.8/15.9922 GB (54.8%)

Time taken: 3.6 sec. A: 7.43 GB, R: 7.94 GB, Sys: 9.3/15.9922 GB (57.9%)
```

**Comment by Haoming02** [comment]
I think this does make sense though, as the prompt chunks are [stacked](https://github.com/Haoming02/sd-webui-forge-classic/blob/1.9/modules/sd_hijack_clip.py#L256) in the end. Meaning the longer the prompt is, the more calculation needs to be done.

As for the CPU being the bottleneck, this is probably because the **Text Encoder** is ran on CPU by default. 

Try adding `--clip-in-gpu` and see if it helps.

**Comment by Zotikus1001** [comment]
I did have --clip-in-gpu already. Its something that happens between or during each generation step most likely.

Btw, the graph above shows the long prompt gen, and the smaller prompt gen. Not gen and hiresfix.

The graph is from MSI Afterburner (set update interval to 100ms instead of 1000 to better see whats going on)

**Comment by Haoming02** [comment]
After digging through the codes, I think the `cond` produced by **CLIP** is kept on the `intermediate_device`...?

And according to [this line](https://github.com/Haoming02/sd-webui-forge-classic/blob/classic/ldm_patched/modules/model_management.py#L718), it is still **CPU**.

Could you try removing `--always-high-vram` and adding `--always-gpu` (keep the `--clip-in-gpu`), to see if that helps?

**Comment by Zotikus1001** [comment]
Nope. Still same CPU bottleneck issue. 

I did also try removing fast fp16 and the bottleneck seems almost gone. but its also slower since its not using fp16 accumulation anymore, so maybe its that the cpu bottleneck wouldnt show anymore, since its slower and not a solution.

I also tried removing sage, and no help.

**Comment by Haoming02** [comment]
Can you do the same test on ComfyUI?

**Comment by Zotikus1001** [comment]
Sure. I'll give it a go today.

**Comment by Zotikus1001** [comment]
So.. 

Clip in GPU.

My findings: 

Comfy with pytorch attention = **_no bottleneck_** (full gpu usage)
Comfy with sage attention = cpu bottleneck

forge-classic with **pytorch sdp attention** = **_no bottleneck_** (full gpu usage)
[Time taken: 6.8 sec | 4.73it/s]

forge-classic with **sage attention** = cpu bottleneck
[Time taken: 8.0 sec | 3.94it/s]

forge-classic with **sage attention** + **--fast-fp16** = cpu bottleneck
[Time taken: 8.0 sec | 3.97it/s]

forge-classic with **sage attention** + **--sageattn2-api auto** = cpu bottleneck
[Time taken: 8.1 sec | 3.91it/s]

forge-classic with **sage attention** + **--sageattn2-api auto** + **--fast-fp16** = cpu bottleneck
[Time taken: 8.0 sec | 3.97it/s]

forge-classic with **xformers attention** = cpu bottleneck
[Time taken: 7.9 sec | 4.01it/s]


After all that, I found something else:

If you use huge prompts, but the exact same on positive and negative, you get full gpu usage (no bottleneck). 
If you delete a few letters or words from any prompt, its fine, but if you delete a full line, bottleneck is back. (doesnt matter where you change it, start, middle or end of prompts)
So, can the issue be in the cfg calculations? or something related to it that happens on every step/during step?

**Comment by Haoming02** [comment]
Take this with a giant grain of salt...

<details>
<summary>ChatGPT (Free Tier)</summary>

The Sage kernel here does a lot more host-side work than PyTorch SDP does.

### Where the CPU can get involved
Looking at the code:
```py
torch.nn.functional.pad(q, ...)
```
This is a CPU op unless the tensors are already contiguous on GPU (but even then, padding might trigger an extra kernel).

<hr>

```py
km = k.mean(dim=seq_dim, keepdim=True)
lse_correction = torch.matmul(...)
```
This is at least one extra reduction + matmul per attention call, which PyTorch SDP does not do. If your sequence length is large (long prompts), this cost scales linearly.

<hr>

```
q_int8, q_scale, k_int8, k_scale = per_thread_int8_triton(...)
v_fp8, v_scale, vm = per_channel_fp8(...)
```
- `per_thread_int8_triton` is a Triton kernel — faster than pure Python, but still a separate GPU launch with its own host-side dispatch.
- `per_channel_fp8` same deal — more launches, more CPU overhead.

<hr>

Instead of one fused `QKᵀ→softmax→V` multiply like PyTorch SDP, Sage:

Runs separate GPU kernels for quantization, scaling, mean subtraction. Then finally launches the main `_qattn_sm89` kernel.

Each launch has a Python → CUDA dispatch cost, which becomes noticeable when per-step work is small (e.g., with shorter sequences), but gets worse when the sequence is longer because the quantization/masking steps scale with sequence length.

<hr>

All the `tensor_layout` handling and `o[..., :head_dim_og]` slicing are lightweight, but they add up if repeated at every step.

`Long prompts → larger qo_len / kv_len → quantization & mean subtraction` steps have more data to chew through.

These steps aren’t fused into the main matmul/softmax kernel, so you pay:

- Python function call overhead.
- Separate GPU kernel dispatches (which involve CPU scheduling).
- Additional memory reads/writes for intermediate tensors.

PyTorch’s SDP avoids this by doing everything in one optimized GPU kernel.

## TL;DR

There’s no obvious “oops, we accidentally run on CPU” in the sense of doing the matmul on CPU —
but Sage does extra host-side orchestration and pre-processing (padding, smooth_k, quantization) that requires multiple kernel launches, so the GPU is idle more often, which feels like a CPU bottleneck in profiling.

</details>

**Comment by Haoming02** [comment]
Anyway, can you open the same issue with the findings on the ComfyUI repo?

If they can make improvements on this, then I can ~~steal~~ reference :tm: their codes here as well :)

**Comment by Zotikus1001** [comment]
I used the sage node from https://github.com/kijai/ComfyUI-KJNodes

I don't think comfy has sage support by default? So maybe not the place for it?

**Comment by Haoming02** [comment]
ComfyUI does have a `sageattention` flag, though IDK how they implement when to use it...

https://github.com/comfyanonymous/ComfyUI/blob/v0.3.50/comfy/cli_args.py#L113

**Comment by Zotikus1001** [comment]
With your tips and some AI help, I managed to fix it:

```py
if isSage2 and args.sageattn2_api is not SageAttentionAPIs.Automatic:
    from functools import partial
    from sageattention import sageattn_qk_int8_pv_fp16_triton, sageattn_qk_int8_pv_fp16_cuda, sageattn_qk_int8_pv_fp8_cuda

    if args.sageattn2_api is SageAttentionAPIs.Triton16:
        sageattn = partial(sageattn_qk_int8_pv_fp16_triton, quantization_backend="cuda")
    elif args.sageattn2_api is SageAttentionAPIs.CUDA16:
        # FIXED: Use supported fp16+fp32 (fp16 operations, fp32 accumulation - perfect for --fast-fp16)
        sageattn = partial(sageattn_qk_int8_pv_fp16_cuda, qk_quant_gran="per_warp", pv_accum_dtype="fp16+fp32")
    elif args.sageattn2_api is SageAttentionAPIs.CUDA8:
        # OPTIMIZED: per_warp + fp32 accumulation 
        sageattn = partial(sageattn_qk_int8_pv_fp8_cuda, qk_quant_gran="per_warp", pv_accum_dtype="fp32+fp32")
else:
    # AUTO MODE: Optimal for FP16 models + --fast-fp16
    if isSage2:
        from functools import partial
        from sageattention import sageattn_qk_int8_pv_fp16_cuda
        # PERFECT: fp16 operations with fp32 accumulation (best speed + stability for your setup)
        sageattn = partial(sageattn_qk_int8_pv_fp16_cuda, qk_quant_gran="per_warp", pv_accum_dtype="fp16+fp32")
        print("Using optimized SageAttention (CUDA fp16 per_warp + mixed precision accumulation)")


def attention_sage(q, k, v, heads, mask=None):
    """
    Reference: https://github.com/comfyanonymous/ComfyUI/blob/v0.3.13/comfy/ldm/modules/attention.py#L472
    Edited by. Haoming02
    OPTIMIZED: Simplified with per_warp quantization and proper tensor handling
    """

    b, _, dim_head = q.shape
    dim_head //= heads

    if (isSage2 and dim_head > 128) or ((not isSage2) and (dim_head not in (64, 96, 128))):
        if model_management.xformers_enabled():
            return attention_xformers(q, k, v, heads, mask)
        else:
            return attention_pytorch(q, k, v, heads, mask)

    # OPTIMIZATION: Ensure contiguous memory layout for optimal SageAttention performance
    q, k, v = map(
        lambda t: t.view(b, -1, heads, dim_head).contiguous(),
        (q, k, v),
    )

    assert mask is None

    # Use optimized SageAttention with per_warp quantization
    out = sageattn(q, k, v, is_causal=False, tensor_layout="NHD")
    return out.reshape(b, -1, heads * dim_head)
```

Now, idk if theres any drawbacks to this as I'm a noob and its tailored to my launch args:

> set COMMANDLINE_ARGS=
>  --uv ^
>  --clip-in-gpu ^
>  --sage ^
>  --sageattn2-api auto ^
>  --fast-fp16 ^
>  --xformers ^
>  --gpu-device-id 0

But I'm guessing you can implement if you deem it has value :)

**Comment by Haoming02** [comment]
Hmm, so the only difference is basically the `SageAttentionAPIs` parts.

Maybe I should open the options more for the users...

**Comment by Zotikus1001** [comment]
Results did change a tiny bit using same seed.

And it didnt completely fix the speed, but its now much better and closer to 99% gpu usage.

**Comment by Haoming02** [comment]
Hello, can you update to the latest commit, and try the following args:

```bash
--sage2-function fp16_cuda --sage-quant-gran per_warp --sage-accum-dtype fp16+fp32
```

see if the speed is as good as your changes?

**Comment by Zotikus1001** [comment]
I can confirm it works :) Same or better speed :D

**Comment by Haoming02** [comment]
Aye, I'll close this Issue then~
